### ISSUE TITLE
[prefer-for-of] Ignore `Symbol.iterator` and `Symbol.asyncIterator` methods

### ISSUE THREAD

**Issue by bbrk24** [typescript-eslint/typescript-eslint#4175]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [ ] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/TROUBLESHOOTING.md) and my problem is not listed. -- That link 404s, so I can't.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-for-of": "error"
  }
}
```

```TS
class MyArray<T> implements ArrayLike<T>, Iterable<T | undefined> {
    length = 0;
    [index: number]: T;

    *[Symbol.iterator](): IterableIterator<T | undefined> {
        for (let i = 0; i < this.length; ++i) {
            yield this[i];
        }
    }
}
```

**Expected Result**

To avoid infinite recursion, `prefer-for-of` should not recommend using a for-of loop on `this` inside the `Symbol.iterator` method.

**Actual Result**

The entire three-line for loop errors, saying ``Expected a `for-of` loop instead of a `for` loop with this simple iteration.`` Fortunately, eslint does not "fix" the error when given the `--fix` flag.

**Additional Info**

The same error happens in `Symbol.asyncIterator`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.4.0` |
| `@typescript-eslint/parser`        | `5.4.0` |
| `TypeScript`                       | `4.4.4` |
| `ESLint`                           | `8.2.0` |
| `node`                             | `14.17.0` |

**Comment by bbrk24** [comment]
Relatedly, I think this is the quickest I've ever caused a stack overflow:

```js
class MyArray {
    *[Symbol.iterator]() {
        for (let el of this) {
            yield el;
        }
    }
}

let arr = new MyArray();
for (let _ of arr) {}
```

**Comment by armano2** [comment]
we should actually do not error when `this[i]` is used, and context should not mater in this case
